jobclient/src/main/java/com/twosigma/cook/jobclient/Constraint.java

```diff
@@ -0,0 +1,170 @@
+        EQUALS("EQUALS");
+
+        Operator(String name) {
+        }
+
+        /**
+         * Returns a constraint by "applying" this operator to the given attribute and value.
+         *
+         * @param attribute The constraint attribute
+         * @param value     The constraint value
+         * @return a constraint.
+         */
+        public Constraint apply(String attribute, String value) {
+            return new Constraint(attribute, this, value);
+        }
+
+        /**
+         * Parse an operator from its string representation.
+         *
+         * @param op specifies a string representation of operator.
+         * @return an operator for the specified name.
+         */
+        public static Constraint.Operator fromString(String op) {
+            return Enum.valueOf(Constraint.Operator.class, op.trim().toUpperCase());
+        }
+    }
+
+    private String _attribute;
+    private Operator _operator;
+    private String _value;
+
+    Constraint(String attribute, Operator operator, String value) {
+        Preconditions.checkNotNull(attribute);
+        Preconditions.checkNotNull(operator);
+        Preconditions.checkNotNull(value);
+        _attribute = attribute;
+        _operator = operator;
+        _value = value;
+    }
+
+    Constraint(String attribute, String operator, String value) {
+        Preconditions.checkNotNull(attribute);
+        Preconditions.checkNotNull(operator);
+        Preconditions.checkNotNull(value);
+        _attribute = attribute;
+        _operator = Operator.fromString(operator);
+        _value = value;
+    }
+
+    /**
+     * Construct a constraint from a {@link JSONArray}.
+     *
+     * The given {@link JSONArray} is assumed to have three strings that
+     * represent attribute, operator, and value of a constraint respectively.
+     *
+     * @param constraint The constraint in {@link JSONArray} representation.
+     */
+    Constraint(JSONArray constraint) {
+        Preconditions.checkNotNull(constraint);
+        Preconditions.checkArgument(constraint.length() == 3);
+        _attribute = constraint.getString(0);
+        _operator = Operator.fromString(constraint.getString(1));
+        _value = constraint.getString(2);
+    }
+
+    /**
+     * @return the attribute of this constraint.
+     */
+    public String getAttribute() {
+        return _attribute;
+    }
+
+    /**
+     * @return the operator of this constraint.
+     */
+    public Operator getOperator() {
+        return _operator;
+    }
+
+    /**
+     * @return the value of this constraint.
+     */
+    public String getValue() {
+        return _value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        Constraint that = (Constraint) o;
+
+        if (_attribute != null ? !_attribute.equals(that._attribute) : that._attribute != null) return false;
```

Comment:
It took me a really long time to grok this, would you mind changing to strict boolean logic. I *think* this is the correct switch:

```java
if (_attribute != null && that._attribute == null) return false;
if (_attribute == null && that._attribute != null) return false;
if (_attribute != null && !_attribute.equals(that._attribute)) return false;
```

I know it is less concise but this makes it much easier to verify that the checks are correct.

Reply:
This is a generated code. Not written by me.

Reply:
intellij is smarter than me. I'd prefer switching it to the boolean logic but knowing it is generated is comforting. @shamsimam what do you think?

Reply:
I like that @wyegelwel version is more readable, but I'm good either way.

Reply:
Maybe in the future, machine/AI generated codes will be completely human unreadable and we have to just trust them. I will just keep as it is.

Reply:
This logic looks like the inverse of [Objects.equals](https://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#equals-java.lang.Object-java.lang.Object-); i.e., it would be much simpler to write this:
```java
if (!java.util.Objects.equals(_attribute, that._attribute)) return false;
```
I'm guessing that the code generator was written before Java 7, which is why they generate this complicated ternary construct instead of using the library. It's too bad that they haven't updated the templates. If you want to make this method more readable, I recommend delegating to the `Objects` static helpers. However, I'm OK with leaving the auto-generated code too—I totally understand being hesitant to "improve" working code when really we're better off just re-generating the methods if we ever update the class again.

Reply:
updated.